### ISSUE TITLE
Wrong solution of linear systems with Armadillo

### ISSUE THREAD

**Issue by valentjn** [SGpp/SGpp#58]
In GitLab by @valentjn on May 15, 2018, 12:23

After updating to Ubuntu 18.04 LTS, the solutions of linear systems in SG++, which is for example needed when interpolating, may be wrong. Necessary for the bug is that all of the following conditions are met:

* The matrix of the linear system is a band matrix, e.g., when interpolating with nodal B-splines.
* The library Armadillo [1] is used for the solution of the linear system.
* The installed Armadillo version is 8.300, 8.400.0, or 8.500.0. This is the case for at least Ubuntu 18.04 LTS (not for 17.10). Whether 18.10 will be affected depends or not depends on whether it will include one of these versions. The bug is fixed in versions >= 8.500.1.

You can check if your Armadillo version is affected by compiling the attached example (by @MichaelReh) and running
```
$ g++ ArmadilloBug.cpp -larmadillo -o ArmadilloBug
$ ./ArmadilloBug
 ||Ax-b|| = 0.0767206
```
The residual should be way less (< 10^(-6)) if the bug is not present.

All credit goes to @MichaelReh for finding the bug!

[1] http://arma.sourceforge.net/

[ArmadilloBug.cpp](/uploads/00c9ac45bdaf0e091f801ab68921943a/ArmadilloBug.cpp)
[A.mat](/uploads/a5d7b354da56dd382d1df249f84dd745/A.mat)

**Comment by valentjn** [comment]
In GitLab by @valentjn on May 15, 2018, 12:24

The bug has been reproduced by @MichaelReh and me.

**Comment by valentjn** [comment]
In GitLab by @valentjn on May 15, 2018, 12:28

Conrad Sanderson, the author of Armadillo, has fixed this in version 8.500.1 (report by @MichaelReh), see [1]. To fix this on your machine, you have to either upgrade or downgrade your version to < 8.300. 8.500.1 is currently not in the Ubuntu repositories.

We will have to see if Ubuntu 18.10 will include the new version.

I think we should close this as invalid as this is not a SG++ bug. However, I suppose it's reasonable to keep this bug report for the record (I heard that @MichaelReh spent three days searching for a bug in his code...).

[1] https://github.com/conradsnicta/armadillo-code/issues/76

**Comment by valentjn** [comment]
In GitLab by @valentjn on May 15, 2018, 12:28

closed
